### PR TITLE
Mark APIs that are "private alpha release" as such.

### LINKED ISSUE #435: Label 'private alpha' features in the code as such.
Cloud Bigtable features (for example snapshots) they go through a release processs (think *alpha* -> *beta* -> *GA*).  At some point we need to support the feature in the Bigtable Client, but we have to let the users know about the state of the feature clearly.  This bug proposes marking these features using both documentation and compiler diagnostics.

### Documentation

We propose grouping all APIs that are not GA in a Doxygen [group](https://www.stack.nl/~dimitri/doxygen/manual/grouping.html), the documentation for the group would include a `@warning` directive to let users know what is going on.

Additionally, each function that exposes a non-GA API should carry the same warning.

### Diagnostics

We propose using the `[[deprecated]]` attribute to label all non-GA functions as `[[deprecated("private alpha release")]]` or `[[deprecated("beta release")]]` or whatever the state of the underlying Cloud Bigtable API is.  Compilers issue a diagnostic (a warning) when such APIs are used (not when they are compiled).

### Downsides

If the application developers have turned on `-Werror` (turn all warnings into errors), their code would not compile when using the code that is not GA.  We will offer helper headers to disable the warnings only in the portions of their code that use this functions for the most common compilers.

### PR DESCRIPTION
**Pull request by coryan**
This fixes #435.  It adds Doxygen warnings to the member functions that access those portions of Cloud Bigtable that are not GA (currently Snapshots).  It also adds compiler warnings if the application developer uses those functions.  And some helper headers to remove the warnings when the application developer really wants to use the APIs.

### CODE REVIEW

**Comment by coryan** [comment]
I forgot to include a link to the generated Doxygen docs:

https://coryan.github.io/google-cloud-cpp/latest/bigtable/classbigtable_1_1v0_1_1TableAdmin.html#pub-methods

https://coryan.github.io/google-cloud-cpp/latest/bigtable/classbigtable_1_1v0_1_1TableAdmin.html#af78b10d08e7326197167df77342f3eee

**Comment by coryan** [comment]
@sduskis @ghaisa @mbrukman I think you folks need to speak out about how this affects the documentation and user experience for pre-release APIs.

File: bigtable/client/internal/wdeprecated_as_warnings.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file is special, it should not have include guards or #pragma once
+// directives
+
+#if defined(__clang__)
+#pragma clang diagnostic warning "-Wdeprecated-declarations"
```
Comment (mbrukman) [inline bigtable/client/internal/wdeprecated_as_warnings.h]: I understand the file name here is named after the flag, but `wdeprecated_as_warnings.h` still reads like a typo to me. Can we drop the leading `w`? Or call it something like `warnings_for_deprecations.h` or similar?

File: bigtable/client/internal/port_platform.h [hunk 2]
```diff
@@ -68,4 +68,10 @@
 #endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 // clang-format on
 
+#ifndef GOOGLE_CLOUD_CPP_DISABLE_PRIVATE_ALPHA_WARNINGS
+#define GOOGLE_CLOUD_CPP_PRIVATE_ALPHA [[deprecated("private alpha release")]]
```
Comment (mbrukman) [inline bigtable/client/internal/port_platform.h]: Curious why alpha features are marked as "deprecated"? Is `@warning` not sufficient for this? Will the user see the word "deprecated" in the output, or just see the word "warning"?

File: bigtable/client/internal/wdeprecated_as_warnings.h (hunk 1 above)
Comment (coryan) [reply bigtable/client/internal/wdeprecated_as_warnings.h]: `diagnostics_deprecated_as_warnings` ?  That creates a sensible pattern for other diagnostics.

File: bigtable/client/internal/port_platform.h (hunk 2 above)
Comment (coryan) [reply bigtable/client/internal/port_platform.h]: The only C++ annotation that creates a warnings is the `[[deprecated]]` annotation.  It is ugly, and it is one of the reasons I wanted you and @ghaisa to take a look.

Comment (mbrukman) [reply bigtable/client/internal/port_platform.h]: Will the literal word "deprecated" appear anywhere in the output (e.g., compiler-generated warning messages), or will it mark these APIs in any way as "deprecated" anywhere (e.g., IDE) other than that C++ macro? Or will users only see the literal warning that you add in the macro after the `deprecated` keyword?

**Comment by coryan** [comment]
@mbrukman @sduskis I have removed the `[[deprecated]]` attributes and the support code for them.  We can have a separate PR for them once if we agree they are worthwhile.

**Pull request merged**